File: src/main/java/com/raychenon/leetcode/MaximumDistanceToClosestPerson.java
package com.raychenon.leetcode;

/**
 * User: raychenon
 * Date: 2019-11-22
 * https://leetcode.com/problems/maximize-distance-to-closest-person
 */
public class MaximumDistanceToClosestPerson {

    public int maxDistToClosest(int[] seats) {
        int N = seats.length;
        int left = -1, right = 0;
        int ans = 0;

        for (int i = 0; i < N; ++i) {
            if (seats[i] == 1) {
                left = i;
            } else {
                while (right < N && seats[right] == 0 || right < i) {
                    ++right;
                }

                int l = left == -1 ? N : i - left;
                int r = right == N ? N : right - i;
                ans = Math.max(ans, Math.min(l, r));
            }
        }

        return ans;
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/MaximumDistanceToClosestPerson.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 2019-11-23
 * https://leetcode.com/problems/maximize-distance-to-closest-person
 */
object MaximumDistanceToClosestPerson {

    /**
     * Use 2 pointers method: left and right
     * Time complexity: O(n)
     * Space complexity: O(1)
     *
     * @param seats
     * @return the maximum distance to closest person.
     */
    fun maxDistToClosest(seats: IntArray): Int {
        val N = seats.size
        var left = -1
        var right = 0
        var ans = 0

        for (i in 0 until N) {
            if (seats.get(i) == 1)
                left = i
            else {
                while (right < N && seats.get(right) == 0 || right < i) {
                    ++right;
                }
                var l = if (left == -1) N else i - left
                var r = if (right == N) N else right - 1
                ans = Math.max(ans, Math.min(l, r))
            }
        }

        return ans

    }
    
}